internal/apidiff/apidiff.go

```
@@ -0,0 +1,255 @@
+
+		// Create apidiff base from repo remote HEAD.
+		base, err := writeBase(baseModDir, imp, pkg)
+		if err != nil {
+			errs[imp] = err
+			continue
+		}
+
+		// Diff the current checked out change against remote HEAD base.
+		out, err := diff(modDir, imp, pkg, base)
+		if err != nil {
+			errs[imp] = err
+			continue
+		}
+
+		if out != "" {
+			diffs[imp] = out
+		}
+	}
+}
+
+func writeBase(baseModDir, imp, pkg string) (string, error) {
+	if err := cd(baseModDir); err != nil {
+		return "", err
+	}
+
+	base := path.Join(baseModDir, "pkg.master")
+	out, err := execE("apidiff", "-w", base, imp)
+	if err != nil && !isSubModErr(out) {
+		return "", err
+	}
+
+	// If there was an issue with loading a submodule, change into that
+	// submodule directory and try again.
+	if isSubModErr(out) {
+		parent := manualParent(imp, manifest)
+		if parent == pkg {
+			return "", fmt.Errorf("unable to find parent module for %q", imp)
+		}
+		if err := cd(parent); err != nil {
+			return "", err
+		}
+		out, err := execE("apidiff", "-w", base, imp)
+		if err != nil {
+			return "", fmt.Errorf("%s: %s", err, out)
+		}
+	}
+	return base, nil
+}
+
+func diff(modDir, imp, pkg, base string) (string, error) {
+	if err := cd(modDir); err != nil {
+		return "", err
+	}
+	out, err := execE("apidiff", "-incompatible", base, imp)
+	if err != nil && !isSubModErr(out) {
+		return "", err
+	}
+	if isSubModErr(out) {
+		parent := manualParent(imp, manifest)
+		if parent == pkg {
+			return "", fmt.Errorf("unable to find parent module for %q", imp)
+		}
+		if err := cd(parent); err != nil {
+			return "", err
+		}
+		out, err = execE("apidiff", "-w", base, imp)
+		if err != nil {
+			return "", fmt.Errorf("%s: %s", err, out)
+		}
+	}
+
+	return out, err
+}
+
+func shouldRun() {
+	head := mustExec("git", "log", "-1")
+	if strings.Contains(head, "BREAKING_CHANGE_ACCEPTABLE") {
+		os.Exit(0)
+	}
+}
+
+func loadManifest() {
+	f, err := os.Open(repoMetadataPath)
+	if err != nil {
+		log.Fatalln(err)
```

nit: for this and other error could we return the error and only os.Exit from the main function?

---

Done!